api/v1/_uploadPublicFile/PKPUploadPublicFileHandler.inc.php

```diff
@@ -0,0 +1,189 @@
+	 * @copydoc PKPHandler::authorize
+	 */
+	public function authorize($request, &$args, $roleAssignments) {
+		import('lib.pkp.classes.security.authorization.PolicySet');
+		$rolePolicy = new PolicySet(COMBINING_PERMIT_OVERRIDES);
+
+		import('lib.pkp.classes.security.authorization.RoleBasedHandlerOperationPolicy');
+		foreach($roleAssignments as $role => $operations) {
+			$rolePolicy->addPolicy(new RoleBasedHandlerOperationPolicy($request, $role, $operations));
+		}
+		$this->addPolicy($rolePolicy);
+
+		return parent::authorize($request, $args, $roleAssignments);
+	}
+
+	/**
+	 * A helper method which adds the necessary response headers to allow
+	 * file uploads
+	 *
+	 * @param $response Response object
+	 * @return Response
+	 */
+	private function getResponse($response) {
+		return $response->withHeader('Access-Control-Allow-Headers', 'Content-Type, X-Requested-With, X-PINGOTHER, X-File-Name, Cache-Control');
+	}
+
+	/**
+	 * Upload a requested file
+	 *
+	 * @param $slimRequest Request Slim request object
+	 * @param $response Response object
+	 * @param $args array arguments
+	 * @return Response
+	 */
+	public function uploadFile($slimRequest, $response, $args) {
+		$request = $this->getRequest();
+
+		if (empty($_FILES) || empty($_FILES['file'])) {
+			return $response->withStatus(400)->withJsonError('api.temporaryFiles.400.noUpload');
+		}
+
+		$siteDir = Core::getBaseDir() . '/' . Config::getVar('files', 'public_files_dir') . '/site';
+
+		if (!file_exists($siteDir) || !is_writeable($siteDir)) {
+			return $response->withStatus(500)->withJsonError('api.publicFiles.500.badFilesDir');
+		}
+		$userDir = $siteDir . '/images/' . $request->getUser()->getUsername();
+		$isUserAllowed = true;
+		$allowedDirSize = 25000 * 1024; // 25mb
+		$allowedFileTypes = ['gif', 'jpg', 'png'];
+
+		HookRegistry::call('API::uploadPublicFile::permissions', [
+			&$userDir,
+			&$isUserAllowed,
+			&$allowedDirSize,
+			&$allowedFileTypes,
+			$request,
+			$this->getAuthorizedContextObject(ASSOC_TYPE_USER_ROLES),
+		]);
+
+		// Allow plugins to control who can upload files
+		if (!$isUserAllowed) {
+			return $response->withStatus(403)->withJsonError('api.publicFiles.403.unauthorized');
+		}
+
+		// Don't allow user to exceed the alotted space in their public directory
+		$currentSize = 0;
+		if ($allowedDirSize > 0 && file_exists($userDir)) {
+			foreach (new RecursiveIteratorIterator(new RecursiveDirectoryIterator($userDir, FilesystemIterator::SKIP_DOTS)) as $object) {
+				$currentSize += $object->getSize();
+			}
+		}
+		if (($currentSize + $_FILES['file']['size']) > $allowedDirSize) {
+			return $response->withStatus(413)->withJsonError('api.publicFiles.413.noDirSpace', [
+				'fileUploadSize' => ceil($_FILES['file']['size']/1024),
+				'dirSizeLeft' => ceil(($allowedDirSize - $currentSize)/1024),
+			]);
+		}
+
+		import('lib.pkp.classes.file.FileManager');
+		$fileManager = new FileManager();
+		$filename = $fileManager->getUploadedFileName('file');
```

Comment:
This gets sourced from `$_FILES['abc']['name']`, which comes from the client computer and should be considered untrustworthy. I'd suggest passing it through a fairly limiting regexp before accepting it.

Reply:
(I could even be convinced that we should assign our own filenames rather than taking them from the client; image uploads from untrusted users are sometimes misrepresented as successful hacks to the system, and I think the filename is part of what gives sysadmins the impression of validity of the "hack".)

Reply:
Convert spaces and underscores to `-` and strip out everything but alphanums, `-` and `.`: https://github.com/pkp/pkp-lib/commit/f86f3f89be82b0d1d3c54e3aefb7307fb66db31d